db/compaction/compaction_job_internal.h

```diff
@@ -0,0 +1,395 @@
+        files_to_cut_for_ttl(std::move(state.files_to_cut_for_ttl)),
+        cur_files_to_cut_for_ttl(state.cur_files_to_cut_for_ttl),
+        next_files_to_cut_for_ttl(state.next_files_to_cut_for_ttl),
+        total_bytes(state.total_bytes),
+        num_output_records(state.num_output_records),
+        compaction_job_stats(std::move(state.compaction_job_stats)),
+        approx_size(state.approx_size),
+        grandparent_index(state.grandparent_index),
+        overlapped_bytes(state.overlapped_bytes),
+        seen_key(state.seen_key),
+        sub_job_id(state.sub_job_id),
+        notify_on_subcompaction_completion(
+            state.notify_on_subcompaction_completion),
+        compaction_outputs_(std::move(state.compaction_outputs_)),
+        preclude_bottommost_outputs_(
+            std::move(state.preclude_bottommost_outputs_)),
+        is_precluded_bottommost_(state.is_precluded_bottommost_),
+        has_precluded_bottommost_outputs_(
+            state.has_precluded_bottommost_outputs_) {
+    current_outputs_ = is_precluded_bottommost_ ? &preclude_bottommost_outputs_
+                                                : &compaction_outputs_;
+  }
+
+  // Adds the key and value to the builder
+  // If paranoid is true, adds the key-value to the paranoid hash
+  Status AddToBuilder(const Slice& key, const Slice& value) {
+    Status s = current_outputs_->AddToBuilder(key, value);
+    num_output_records++;
+    return s;
+  }
+
+  bool HasBuilder() const { return current_outputs_->builder != nullptr; }
+
+  bool HasPrecludedBottommostOutputs() const {
+    return has_precluded_bottommost_outputs_;
+  }
+
+  bool HasOutput() { return !current_outputs_->outputs.empty(); }
+
+  void FillFilesToCutForTtl();
+
+  // Returns true iff we should stop building the current output
+  // before processing "internal_key".
+  bool ShouldStopBefore(const Slice& internal_key, uint64_t curr_file_size);
+
+  Status ProcessOutFlowIfNeeded(const Slice& key, const Slice& value) {
+    return current_outputs_->ProcessOutFlowIfNeeded(key, value);
+  }
+
+  bool IsPrecludedBottommost() const { return is_precluded_bottommost_; }
+
+  void OverrideCurrentOutputs(bool value) {
+    is_precluded_bottommost_ = value;
+    current_outputs_ = is_precluded_bottommost_ ? &preclude_bottommost_outputs_
+                                                : &compaction_outputs_;
+  }
+
+  bool SetPrecludedBottommost(const CompactionIterator* iter);
+
+  void AddOutput(FileMetaData& meta, bool enable_hash);
+
+  void AddOutput(FileMetaData& meta, uint64_t paranoid_hash);
+
+  Status AddRangeTombstones(CompactionRangeDelAggregator* range_del_agg,
+                            CompactionIterationStats* range_del_out_stats,
+                            SequenceNumber earliest_snapshot,
+                            const Slice* next_table_min_key) {
+    auto cfd = compaction->column_family_data();
+    return current_outputs_->AddRangeTombstones(
```

Comment:
I didn't spend time thinking about where we should put range tombstones. Can you add a comment explaining the reason for the current implementation?

Reply:
Added.